extensions/amp-sidebar/1.0/sidebar.js

```diff
@@ -0,0 +1,241 @@
+  return valueRef;
+}
+/**
+ * @param {!Element} element
+ * @param {!Object<string, *>} styles
+ */
+function safelySetStyles(element, styles) {
+  setStyles(element, assertDoesNotContainDisplay(styles));
+}
+
+/**
+ * @param {!SidebarDef.Props} props
+ * @param {{current: (!SidebarDef.SidebarApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function SidebarWithRef(
+  {
+    as: Comp = 'div',
+    side = 'left',
+    onBeforeOpen,
+    onAfterClose,
+    children,
+    ...rest
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `opened` plays the animation.
+  const initialOpen = false;
+  const [mounted, setMounted] = useState(initialOpen);
+  const [opened, setOpened] = useState(initialOpen);
+  const classes = useStyles();
+  const sidebarRef = useRef();
+  const maskRef = useRef();
+
+  // We are using refs here to refer to common strings, objects, and functions used.
+  // This is because they are needed within `useEffect` calls below (but are not depended for triggering)
+  // We use `useValueRef` for props that might change (user-controlled)
+  const onBeforeOpenRef = useValueRef(onBeforeOpen);
+  const onAfterCloseRef = useValueRef(onAfterClose);
+
+  const open = useCallback(() => {
+    if (onBeforeOpenRef.current) {
+      onBeforeOpenRef.current();
+    }
+    setMounted(true);
+    setOpened(true);
+  }, [onBeforeOpenRef]);
+  const close = useCallback(() => setOpened(false), []);
+  const toggle = useCallback(() => (opened ? close() : open()), [
+    opened,
+    open,
+    close,
+  ]);
+
+  useImperativeHandle(
+    ref,
+    () =>
+      /** @type {!SidebarDef.SidebarApi} */ ({
+        open,
+        close,
+        toggle,
+      }),
+    [open, close, toggle]
+  );
+
+  useLayoutEffect(() => {
+    const sidebarElement = sidebarRef.current;
+    const maskElement = maskRef.current;
+    if (!sidebarElement || !maskElement) {
+      return;
+    }
+
+    let sidebarAnimation;
+    let maskAnimation;
+    // "Make Visible" Animation
+    if (opened) {
+      const postVisibleAnim = () => {
+        safelySetStyles(sidebarElement, ANIMATION_STYLES_SIDEBAR_FINAL);
+        safelySetStyles(maskElement, ANIMATION_STYLES_MASK_FINAL);
+      };
+      if (!sidebarElement.animate || !maskElement.animate) {
+        postVisibleAnim();
+        return;
+      }
+
+      safelySetStyles(
+        sidebarElement,
+        side === 'left'
+          ? ANIMATION_STYLES_SIDEBAR_LEFT_INIT
+          : ANIMATION_STYLES_SIDEBAR_RIGHT_INIT
+      );
+      safelySetStyles(maskElement, ANIMATION_STYLES_MASK_INIT);
+      sidebarAnimation = sidebarElement.animate(
+        side === 'left'
+          ? ANIMATION_KEYFRAMES_SLIDE_IN_LEFT
+          : ANIMATION_KEYFRAMES_SLIDE_IN_RIGHT,
+        {
+          duration: ANIMATION_DURATION,
+          fill: 'both',
```

Comment:
FMI: Why are some of these style objects given constant doc level names and others inlined here?

Reply:
It is a tiny bit wonky, but my thought process was -

`fill` and `direction` are not configurable as they are fundamentally needed for the way the animation works.
`duration` and `ease-in` can be configured to make the animation slower / faster, curvier, and it still works, so thought it made sense to make as a constant.  

I'm not married to it either way so let me know if you feel strongly about it.